Migration of affine schemes

HereAround opened the pull request.

HereAround committed:
Schemes: Move affine schemes to main source

Co-authored-by: Matthias Zach <HechtiDerLachs@users.noreply.github.com>

simonbrandhorst commented:
Tests fail. So this pull requests just reorders code but does not add anything new?

HereAround commented:
The goal is to ship the code for the affine schemes into the main part of Oscar. (And later the more involved parts of schemes, but we have to start somewhere...).

Right now, this PR (actually a draft) is only work in progress of @HechtiDerLachs and myself. Hopefully more tomorrow/in the upcoming days.

Fine print:
- place the code for the affine schemes in the "main" part of Oscar,
- restructure it for easier reading/debugging etc. (I suggest something like we have for the toric varieties),
- add extensive documentation and examples,
- and possibly more that we might spot along the way.

HereAround committed:
Schemes: Add file for easier execution of tests

Co-authored-by: Matthias Zach <HechtiDerLachs@users.noreply.github.com>

HereAround committed:
Schemes: Spread out affine schemes code for better readability

Co-authored-by: Matthias Zach <HechtiDerLachs@users.noreply.github.com>

simonbrandhorst commented on docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:
```diff
@@ -6,229 +6,111 @@ CurrentModule = Oscar
 
+## Methods
+
+With `==`, one can test if two affine schemes are isomorphic.
```
Two subschemes `X` and `Y` of the same ambient affine space `A` are considered equal if and only if the identity morphism of `A`
induces an isomorphism of `X` and `Y`.  This is implemented by `X==Y`. 
For `X` and `Y` in different ambient affine spaces, `X==Y` is always `false`.

HereAround replied:
Thank you. I just added this.

HereAround committed:
Schemes: Code improvement
- Fix minor typos
- Add display method for morphisms of affine schemes
- Improve equality of affine schemes

Co-authored-by: Matthias Zach <HechtiDerLachs@users.noreply.github.com>

HereAround committed:
Schemes: Add documentation for affine schemes

HereAround commented:
@wdecker  @simonbrandhorst  @afkafkafk13  @HechtiDerLachs (and everybody else interested in schemes).

As usual - I do not claim that this is perfect - at least not yet. But I hope it is a good start to achieve:
1. Lifting the schemes from experimental to the main code source of `Oscar`.
2. Providing documentation for the users (such as myself) that want to play with the schemes.

Hence, this PR conducts the following:
- it moves the code for affine schemes from experimental into the main code source.
- It then unpacks this code and organizes it. The idea being that now the code is much easier to read.
- Last but not least, it add excessive documentation including tons of examples.

So at the very least - this PR should give examples on what can (and cannot) currently be done with affine schemes in Oscar. In addition, it might also help (a bit) with the discussion in https://github.com/oscar-system/Oscar.jl/pull/1659 (I appreciate that @afkafkafk marked this other PR as draft. If it got merged before this massive PR here, I would have to go through an involved rebase. Thx.)

Please let me know what you think.

(As @fingolfin pointed out on Slack, the julia master broke ABI compatibility. I believe this is the reason for the failures in the tests with the nightly julia version.)

afkafkafk13 commented on docs/doc.main:
```diff
@@ -147,13 +147,11 @@
-#        "Experimental/non_plane_curves.md",
-         "AlgebraicGeometry/para_rational_curves.md",
+     "Schemes" => [
+        "AlgebraicGeometry/Schemes/GeneralSchemes.md",
```
When I started writing Documentation, @wdecker informed me that we should try to start with an intro.md. Is GeneralSchemes.md this Introduction, without adhering to the naming convention?

HereAround replied:
I was not aware of this naming convention. But yes - the GeneralSchemes.md is supposed to be the starting point. It can be renamed.

afkafkafk13 commented on docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:
````diff
@@ -0,0 +1,142 @@
+```@docs
+Spec(R::MPolyRing, I::MPolyIdeal)
+Spec(R::MPolyRing, U::AbsMPolyMultSet)
+Spec(R::MPolyRing, I::MPolyIdeal, U::AbsMPolyMultSet)
````
I assume that you have reasons for not allowing Spec(R::MPolyRing)? Why do you not allow to create A^n from an MPolyRing directly?

HechtiDerLachs replied:
That is an internal constructor and afaik there is no space for a docstring there.

HereAround replied:
@HechtiDerLachs knows his code better than I do. From what I see, there is no constructor (yet) which simply accepts a polynomial ring. This could be added (maybe in a separate PR - this one is already huge). Alternatively, one can use the `affine_space` command to create the affine space. Feature request?

afkafkafk13 replied:
Yes, this should go into a separate PR, if we want it. It is one of these compatibility features to make life simpler for the scripting user.

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+@doc Markdown.doc"""
+    Spec(R::MPolyRing, U::AbsMPolyMultSet)
+
+Given a polynomial ring ``R`, we can localize that polynomial
```
```suggestion
Given a polynomial ring ``R``, we can localize that polynomial
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    Spec(R::MPolyRing, U::AbsMPolyMultSet)
+
+Given a polynomial ring ``R`, we can localize that polynomial
+ring at a multiplicatively closed subset ``U`` of ``R`. The spectrum
```
```suggestion
ring at a multiplicatively closed subset ``U`` of ``R``. The spectrum
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+of the localized ring ``U^{-1} R`` is computed by this method.
+
+# Examples
+Set ``C`` to be the positive orthant in two dimensions.
```
```suggestion
We create the complement of the  ``y``-axis in the affine plane.
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    Spec(R::MPolyRing, I::MPolyIdeal, U::AbsMPolyMultSet)
+
+We allow to combine quotients and localizations at the same time.
+That is, consider a polynomial ring ``R`, an ideal ``I`` of R and
```
```suggestion
That is, consider a polynomial ring ``R``, an ideal ``I`` of ``R`` and
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+
+We allow to combine quotients and localizations at the same time.
+That is, consider a polynomial ring ``R`, an ideal ``I`` of R and
+a multiplicatively closed subset ``U`` of ``R`. The spectrum of the
```
```suggestion
a multiplicatively closed subset ``U`` of ``R``. The spectrum of the
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+localized ring ``U^{-1} (R/I)`` is computed by this method.
+
+# Examples
+Set ``C`` to be the positive orthant in two dimensions.
```
```suggestion
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+"""
+Spec(R::MPolyRing, I::MPolyIdeal, U::AbsMPolyMultSet) = Spec(MPolyQuoLocalizedRing(R, I, U))
+
+
```
What about the ``Spec(R)`` where ``R`` is an `MPolySomething`?

HereAround replied:
~~Simply not there yet.~~ I refer to @HechtiDerLachs for more details.

I can be a bit more specific. Right now, the ring `R` can be one of the following:
- (Polynomial) ring,
- Quotient of (polynoimal) ring,
- Localization of (polynomial) ring,
- Localization of quotient of (polynomial) ring.

Do you have something more general in mind? Or are you complaining that an implementation is missing for one of these types?

simonbrandhorst replied:
I was just missing the corresponding docstring.

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+@doc Markdown.doc"""
+    affine_space(kk::BRT, n::Int; variable_name="x") where {BRT<:Ring}
+
+The ``n`-dimensional affine space over a ring ``kk`` is created
```
```suggestion
The ``n``-dimensional affine space over a ring ``kk`` is created
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+@doc Markdown.doc"""
+    affine_space(kk::BRT, var_symbols::Vector{Symbol}) where {BRT<:Ring}
+
+This method also creates the ``n`-dimensional affine space over
```
```suggestion
Creates the ``n``-dimensional affine space over
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    affine_space(kk::BRT, var_symbols::Vector{Symbol}) where {BRT<:Ring}
+
+This method also creates the ``n`-dimensional affine space over
+a ring ``kk`, but allows more flexibility in the choice of variable
```
```suggestion
a ring ``kk``, but allows more flexibility in the choice of variable
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+
+
+########################################################
+# (4) StdSpec (needed?)
```
Think that `standard_spec` is for internal use only and should not be part of our official interface. Hence not exported/part of the documentation. As far as I remember it is a relic of the past. (There should be docstrings anyways ofcourse.)

HereAround replied:
Ok. I believe one the notions to discussed. I am pretty sure @HechtiDerLachs has an opinion on the matter.

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+@doc Markdown.doc"""
+    subscheme(X::AbsSpec, f::Vector{<:RingElem})
+
+For an affine spectrum ``X`` and elements ``f1`, ``f2`,
```
```suggestion
For an affine spectrum ``X`` and elements ``f1`, ``f2``,
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    subscheme(X::AbsSpec, f::Vector{<:RingElem})
+
+For an affine spectrum ``X`` and elements ``f1`, ``f2`,
+etc. of the coordinate ring of ``X`, this method computes
```
```suggestion
etc. of the coordinate ring of ``X``, this method computes
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+@Markdown.doc """
+    hypersurface_complement(X::AbsSpec, f::Vector{<:RingElem})
+
+For a scheme ``X = Spec(R)`` and elements ``f1, f2, ... ∈ R``
```
Looks a little wierd in the latex style docu. Do unicode subscripts improve the situation?

simonbrandhorst commented on docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:
````diff
@@ -0,0 +1,142 @@
+### Other attributes
+
+```@docs
+base_ring(X::AbsSpec)
````
Somehow this also includes `base_ring(I::MPolyIdeal)`  and `base_ring(M::PMat)` in the docu. No clue why. Please fix.

simonbrandhorst replied:
Same for codim

HechtiDerLachs replied:
There's little we can do about such artifacts. Sometimes this happens. It's a bug in the julia documenter, I think.

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
```diff
@@ -0,0 +1,275 @@
+@Markdown.doc """
+    defining_ideal(X::AbsSpec{<:Any, <:MPolyRing})
+
+This method return the defining ideal of an affine scheme ``X``.
```
```suggestion
Return the defining ideal of the affine scheme ``X``.
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
```diff
@@ -0,0 +1,275 @@
+@Markdown.doc """
+    strict_modulus(X::AbsSpec)
+
+This method return the strict modulus of an affine scheme ``X``.
```
```suggestion
This method returns the strict modulus of an affine scheme ``X``.
```

simonbrandhorst replied:
No clue what a `strict_modulus` is. For me this means that either we need a self explanatory name or the function is internal. Or that we should work more with morphisms.

simonbrandhorst replied:
So for now I would leave everything out of the interface that is not rigorously mathematically defined and part of the standard literature.

HereAround replied:
I will commit your suggestion now to fix the typo. Let me remove this entry from the documentation in a separate commit shortly.

simonbrandhorst commented on src/Schemes/AffineSchemes/Objects/Properties.jl:
```diff
@@ -0,0 +1,487 @@
+julia> isempty(X)
+false
+
+julia> isempty(EmptyScheme(QQ))
```
Another example could be a subscheme defined by the (1) ideal

HereAround replied:
Thank you. I have added this.

simonbrandhorst commented on src/Schemes/AffineSchemes/Morphisms/Attributes.jl:
```diff
@@ -0,0 +1,260 @@
+@Markdown.doc """
+    graph(f::AbsSpecMor)
+
+This method computes the preimage of a morphism.
```
```suggestion
Return the graph of $f: X \to Y$  as a subscheme of $X \times Y$ as well as the two projections to $X$ and $Y$.
```

simonbrandhorst commented on src/Schemes/AffineSchemes/Morphisms/Methods.jl:
```diff
@@ -0,0 +1,148 @@
+    fiber_product(f::SpecMor{SpecType, SpecType, <:Any}, g::SpecMor{SpecType, SpecType, <:Any}) where {SpecType<:StdSpec}
+
+For morphisms ``f : Y → X`` and ``g : Z → X`` return the fiber
+product ``Y ×ₓ Z`` over ``X`` together with its two canonical
```
```suggestion
product  $Y \times_X Z$ over ``X`` together with its two canonical
```
The unicode looks too wierd to be an improvement.

HereAround replied:
I understand that @HechtiDerLachs uses a keyboard layout that allows him to directly enter these symbols. Some of them (e.g. the element symbol or the mapping arrow symbols above) I found rather nice. But I agree that this here is somewhat small and it is probably easier to read the LaTeX-code.

simonbrandhorst commented on docs/src/AlgebraicGeometry/Schemes/ArchitectureOfAffineSchemes.md:
````diff
@@ -0,0 +1,149 @@
+               }
+```
+Any such morphism has the attributes `domain`, `codomain` and `pullback`.
+A concrete and minimalistic implementation exist for the type `SpecMor`:
````
```suggestion
A concrete and minimalistic implementation exists for the type `SpecMor`:
```

simonbrandhorst reviewed (changes requested):
Thank you for putting so much work into the schemes part. Especially the documentation. 
I think that it is certainly an improvement.  (Though the difference between attributes methods and properties eludes me).
Some comments are included. Most are just typos.

afkafkafk13 commented on docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:
````diff
@@ -0,0 +1,142 @@
+### Intersections
+
+```@docs
+Base.intersect(X::AbsSpec{BRT, <:MPolyRing}, Y::AbsSpec{BRT, <:MPolyRing}) where {BRT<:Ring}
````
This is certainly not the only case, but maybe the only one which already has documentation?

HechtiDerLachs replied:
Which other case should there be? I wouldn't be surprised if this was the only one.

HereAround replied:
There are many (order 10) different `Base.intersect` methods implemented in the affine schemes, depending on whether they are defined from polynomial ring, quotients of polynoimal rings etc. To me, it felt too much to document all of these functions. So I documented the first non-trivial instance. But maybe there is a better way?

afkafkafk13 replied:
For the novice reader, this looks like only being able to supply MPolyRing to it. This should be avoided -- but what is the best way do that?

simonbrandhorst replied:
Define `Base.intersect(X::AbsSpec, Y::AbsSpec)` as not implemented yet and document it.
Then its documentation should show up.

afkafkafk13 commented on docs/src/AlgebraicGeometry/Schemes/AffineSchemes.md:
```diff
@@ -0,0 +1,142 @@
+
+### Comparison
+
+Schemes can be compared based on their `ambient_ring`.
```
this should rather be: 
Comparison of schemes is available for schemes with the same ambient scheme, i.e. as subschemes of the same ambient scheme.

HechtiDerLachs replied:
I don't think there should be an "ambient scheme", really. I know we have `ambient_ring`, but the embedding is not really part of the mathematical data. We should just say that two schemes compare if they have the same `ambient_ring` in the background and refer the user to look that up if necessary.

simonbrandhorst replied:
Maybe things would become more clear if we just declare it as part of the mathematical data,
i.e. that we work with embedded (affine) schemes - that is where the coordinates come from.

afkafkafk13 replied:
Indeed, that is the direction in which I was thinking.

afkafkafk13 commented on experimental/Schemes/Types.jl:
```diff
@@ -702,24 +516,24 @@ end
 
-**Note:** The distinction between the different affine patches of the scheme 
-is made from their hashes. Thus, an affine scheme must not appear more than once 
+**Note:** The distinction between the different affine patches of the scheme
```
This could even be a warning!

afkafkafk13 commented on src/Schemes/AffineSchemes/Morphisms/Attributes.jl:
```diff
@@ -0,0 +1,260 @@
+ x2
+ x3
+
+julia> Y = subscheme(X, x1)
```
Good simple example, but the roles of X and Y are inversed in comparison to the specification. Maybe use X and Z as names instead? (also at other places)

HereAround replied:
Thank you for the pointer. Let me change this.

afkafkafk13 commented on src/Schemes/AffineSchemes/Morphisms/Attributes.jl:
```diff
@@ -0,0 +1,260 @@
+@Markdown.doc """
+    graph(f::AbsSpecMor)
+
+This method computes the preimage of a morphism.
```
Wrong text/Cut&Paste error? graph should not compute a preimage, but a graph, shouldn't it?

HereAround replied:
Yes, that is a cut&paste error. @simonbrandhorst pointed this out above. This should be fixed now.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Types.jl:
```diff
@@ -0,0 +1,81 @@
+  OO::RingType
+  kk::BaseRingType
+
+  function Spec(OO::MPolyQuoLocalizedRing)
```
Comment outside the review: shouldn't we implement a union MPolyAnyRing which consists of all four of the types? This quadruple code fragment looks kind of strange.

HechtiDerLachs replied:
But there is a `base_ring` induced difference, isn't it? I think, having the four spelled out doesn't hurt.

HereAround commented:
Thank you @simonbrandhorst for the many suggestions. Let me work through them.

Concerning properties, attributes and methods of affine schemes:
- Any function which accepts an affine scheme as its only argument and returns a boolean value is a property. Examples include `is_smooth`, `is_singular`, `is_projective`, etc.
- Any function which accepts an affine scheme as its only argument and returns a non-boolean value is an attribute. Examples include `dim`, `codim`, `OO` (or `coordinate_ring`) etc.
- Functions which accept more than one argument and at least one affine scheme are placed in the methods file.

You may wonder what the rational behind this structure is (perse there is no right nor wrong way). My personal motivation dates back about 1.5 years when I started to migrate code from the `homalg_project` (https://github.com/homalg-project) - specifically my `ToricVarietites_project` (https://github.com/homalg-project/ToricVarieties_project) - into Oscar. By mimicking this gap-code, the code for the ToricVarieties in Oscar is now systematically and without exceptions structures in this way. With that being said and making an attempt to aim for a somewhat unified coding approach for algebraic geometry in Oscar (schemes, curves, toric_varieties, tropical geometry,...), I propose to use (or extend as needed) this structure also for the schemes code.

In any case - for me it now very easy to grasp to what extend functionality for affine schemes exists. And, as you pointed out yourself, this includes some auxiliary/helper functions (e.g. the `strict modulus` - I do not know what this is neither), the need/use of which @HechtiDerLachs can certainly explain in large detail. So I see this PR also as a starting point towards a streamlining process of the schemes code.

Either way - on second thought I agree with you. Let me remove those docstrings for now. More shortly.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Types.jl:
```diff
@@ -0,0 +1,81 @@
+    return new{typeof(kk), typeof(OO)}(OO, kk)
+  end
+
+  function Spec(R::Ring)
```
Just Ring and then it is understood as over ZZ. Is this correct or is this just a placeholder for future use?

HechtiDerLachs replied:
This is for future use. Otherwise we would write the type of `ZZ` directly.

HechtiDerLachs replied:
Sorry, I didn't see the full context. Yes, if the user does not specify anything, we interpret the input as a `Spec` over `ZZ`.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Methods.jl:
````diff
@@ -0,0 +1,93 @@
+false
+```
+"""
+function is_non_zero_divisor(f::RingElem, X::AbsSpec{<:Ring, <:MPolyRing})
````
Does this work for Spec Z/4Z[x] and the element 2 or automatically throw an error exception? If not, place a TODO-comment to either secure this against polynomial rings over rings with zerodivisors or handle it properly.

HechtiDerLachs replied:
The generic code is correct. If something is missing down the road, it will complain there. I know Dan complained about me "kicking the can", but I still think it's reasonable, to have the complaint where the problem lays. I don't remember, whether this particular example worked or not. Either way: The implementation is not mandatory for Oscar's ring interface.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
```diff
@@ -0,0 +1,275 @@
+an embedding of ``X`` into ``Spec(R)``. This method returns
+the codimension of the image of this embedding in ``Spec(R)``.
+
+# Examples
```
I would like to see a slightly less trivial example, maybe codim(Spec(K[x,y]/<y>) ?

HereAround replied:
Sure. I assume you meant a closed subscheme, e.g. V(X) so that the codimension is one, i.e. K[x,y]/<x>?

afkafkafk13 replied:
Yes, indeed I wanted K[x,y]/x

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
````diff
@@ -0,0 +1,275 @@
+ideal(x1*x2)
+```
+"""
+@attr defining_ideal(X::AbsSpec{<:Any, <:MPolyRing}) = ideal(OO(X), [zero(OO(X))])
````
(TODO-comment) This should be changed to modulus(OO(X)) for all settings, as soon as we have changed modulus to sensible default settings for Non-Quo-Rings. But for the moment this is ok.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
```diff
@@ -0,0 +1,275 @@
+@Markdown.doc """
+    strict_modulus(X::AbsSpec)
+
+This method return the strict modulus of an affine scheme ``X``.
```
Explain strict modulus, as you are explaining other data.

HereAround replied:
I would love to, but I do not (yet) know what it is. It appears @simonbrandhorst is also at a loss there from what I read above. Please do explain it to us (e.g. on Thursday). Thank you.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Attributes.jl:
```diff
@@ -0,0 +1,275 @@
+# (3) Further attributes for Spec
+########################################################################
+
+strict_modulus(X::Spec) = saturated_ideal(modulus(OO(X)))
```
What is the difference between this an modulus(quotient_ring(OO(X)) (which can be cheaper)?

HechtiDerLachs replied:
I don't remember, to be honest. All this should be streamlined now.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    Spec(R::MPolyRing, U::AbsMPolyMultSet)
+
+Given a polynomial ring ``R`, we can localize that polynomial
+ring at a multiplicatively closed subset ``U`` of ``R`. The spectrum
```
sideremark: Actually in literature it is customary to call the multiplicatively closed set S, not U.

HechtiDerLachs replied:
Wolfram changed this to U so that S could be a ring. We should stick to that.

afkafkafk13 replied:
ok. Fine with me -- as long as it is consistent in the documentation

HereAround replied:
To me the name makes sense. `U` indicates a set with units. Often `R`, `S` are used for rings, so this has the added advantage of leaving `S` free.

afkafkafk13 replied:
It also makes sense to me, I just wanted to point out that in Textbooks one often finds S^-1 R for the localization.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+    standard_spec(X::AbsSpec{<:Any, <:MPolyRing})
+
+For an affine spectrum whose coordinate ring is a polynomial
+ring, this method computes the standard spectrum.
```
What is the standard spectrum? Is this just for internal reasons or is this user facing? In the latter case, the documentation needs to mention what it is and what it is good for.

HechtiDerLachs replied:
This is internal and does not necessarily need to be documented. It turns an arbitrary Spec into one with an `MPolyQuoLocalizedRing` inside.

afkafkafk13 replied:
Then maybe just change the documentation to "this method ensures a standardized internal represenation of a Spec"

HereAround replied:
It was also request by @simonbrandhorst above to remove the documentation for this method (at least for now). I have done so. The `doc_strings` are still there, but are not called in the documentation. So in case this is needed in the future, it will be easy to add this.

HereAround committed:
Schemes: Improve wording of documentation of affine schemes

Co-authored-by: simonbrandhorst <51749255+simonbrandhorst@users.noreply.github.com>

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -0,0 +1,670 @@
+
+For a scheme ``X = Spec(R)`` and elements ``f1, f2, ... ∈ R``
+this returns the open subscheme ``U = Spec(R[f1⁻¹,f2⁻¹, ...]) = X ∖ V(f1, f2, ...)``
+defined by the complement of the vanishing locus ``V(f1, f2, ...)``.
```
What is intended here is X \setminus (V(f1*f2*...), i.e. X \setminus (V(f1) \cup V(f2) \cup ...), not X \setminus (V(f1,f2,...)) = X\setminus (V(f1) \cap V(f2) ...). The syntax using a vector is only intended to pass on a known factorization to avoid repeatedly using factorization.

HechtiDerLachs replied:
We will talk this through and correct it. As it stands here now, it's wrong, but probably hard to guess from the code alone.

HereAround replied:
Ok. Thank you. Let me change that.

afkafkafk13 commented on src/Schemes/AffineSchemes/Objects/Properties.jl:
```diff
@@ -0,0 +1,487 @@
+
+
+# (2.0) For empty schemes and whenever issubset cannot be implemented
+# (2.0) For empty schemes and whenever issubset cannot be implemented
```
duplicate line?

HereAround replied:
Yes - thank you. (Seems I considered this a good idea a couple of days ago. But I agree this is not that good. Removed this and three other such lines just now.)

afkafkafk13 commented on src/Schemes/AffineSchemes/Morphisms/Methods.jl:
```diff
@@ -0,0 +1,148 @@
+@Markdown.doc """
+    simplify(X::AbsSpec{<:Field})
+
+Given an affine scheme ``X`` with coordinate ring ``R = 𝕜[x₁,…,xₙ]/I`` 
```
This changes the ambient scheme and should be documented is this way -- maybe by adding a phrase explaining this on the scheme level.

HechtiDerLachs replied:
To me, this is a reason not to have an "ambient scheme". And the change is intended and necessary, since we want to use the resulting scheme in a new covering.

afkafkafk13 replied:
If we are dealing with embedded schemes, then we can be honest about this change, which is very useful and needed. Right now, we hide the embedding... This needs to be discussed.

afkafkafk13 commented on src/Schemes/AffineSchemes/Morphisms/Constructors.jl:
````diff
@@ -0,0 +1,182 @@
+julia> inclusion_map(Y, X);
+```
+"""
+inclusion_map(X::AbsSpec, Y::AbsSpec) = SpecMor(X, Y, gens(ambient_ring(Y)))  # TODO: Remove
````
Where is the problem about removing this line? Is it used somewhere outside this pull request?

HereAround replied:
I was hoping we could have this PR solemnly for moving the existing code and documenting it. Improving/streamlining it should imho be a separate PR (not only because this PR is already massive as it is). This is why I did not yet touch this/try code optimizations.

HereAround replied:
But it seems we are trying both now... so let me just do it.

afkafkafk13 commented on src/Schemes/AffineSchemes/Morphisms/Constructors.jl:
```diff
@@ -0,0 +1,182 @@
+"""
+function restrict(f::SpecMor, U::AbsSpec, V::AbsSpec; check::Bool=true)
+  if check
+    issubset(U, domain(f)) || error("second argument does not lay in the domain of the map")
```
...does not **lie** in ...

HereAround replied:
Thank you. Fixed.

afkafkafk13 reviewed (commented):
Wow that was a large PR. Great job!
I have a lot of minor remarks, some for you, some for earlier authors. If a comment does not directly affect content of your work and is not 100% trivial, feel free to hand it to the previous author.

HereAround commented:
> Wow that was a large PR. Great job! I have a lot of minor remarks, some for you, some for earlier authors. If a comment does not directly affect content of your work and is not 100% trivial, feel free to hand it to the previous author.

Thank you very much indeed! I am merely the humble person to initiate the migration of this code into the main code source of `Oscar`. The real contributions came from @HechtiDerLachs , which is why I required him to be cc'ed in all commits that migrate this code. The only original piece that I am adding here is to first copy and organize his code and then add examples which hopefully make sense.

HereAround commented:
@simonbrandhorst and @afkafkafk13 Thank you for the many suggestions. I have been able to address most of them.

@simonbrandhorst I squashed your suggestions into https://github.com/oscar-system/Oscar.jl/pull/1652/commits/065adc7773d8d1b49298b874c6da6e34da7e4b1d. Please notice that some of your proposed changes led to failures (e.g. change in `fibre_product` to $Y \times_X Z$ did prevent `Julia` from starting up - it feels the documenter is a bit picky at times). So I had to reverse a few changes to get everything running again.

As requested by @simonbrandhorst, I also removed `standard_spectrum` and `strict_modulus` from `affine_schemes.md`. The `doc_strings` still exist in the source code, and so the documentation can be easily reactivated (once it is clear what these functions do/if they should be communicated to users).

From what I see, the real elephant in the room are the doubled docstrings for `base_ring`, `dim` and `codim`. As I see it, it looks as if some functions do not specify the type of input variables. As a consequence, these functions seem to "believe" that an abstract scheme could be used as input. And so, their docstrings are being displayed at this utterly unexpected position within the affine schemes documentation. The origins are in this case the implementations for rings and tropical geometry in `Oscar` as well as functionality of `Hecke`. I personally think that this issue will be sufficiently complicated to be addressed in a separate PR. If you disagree and want to have this fixed here, then we need to ping the people that could fix this. However, notice that changes in `Hecke` take a couple of days to be migrated to `Oscar`. So the fixes will also likely take some time to be implemented and to take effect.

afkafkafk13 commented:
One more remark (not stopping a merge of PR1652!!): we should make sure that all remarks which came up in this PR, but are actually outside the migration and hence not solved here, end up on a list. We should then quickly go through the list on thursday, discuss the discussion items and check which of the other items could/should be resolved and which can wait.

simonbrandhorst commented:
@HereAround  I sdem to remember that `$` characters are allowed in the docs if it is written like 
```
@doc Markdown.doc"""
    inv(x)

Return $\frac{1}{x}$.
"""
function foo(x)
  return 1//x
end
```
In any case I think we are ready for a merge. I would like to leave out anything controversial
(such as https://github.com/oscar-system/Oscar.jl/issues/1664) to another PR. Including the `base_ring` stuff. 
I will merge when @HechtiDerLachs gives his O.K.

HechtiDerLachs reviewed (approved):
Once the tests pass, we should be good to go.

simonbrandhorst commented:
Tests fail.

HereAround commented:
> @HereAround I sdem to remember that `$` characters are allowed in the docs if it is written like
> 
> ```
> @doc Markdown.doc"""
>     inv(x)
> 
> Return $\frac{1}{x}$.
> """
> function foo(x)
>   return 1//x
> end
> ```
> 
> In any case I think we are ready for a merge. I would like to leave out anything controversial (such as #1664) to another PR. Including the `base_ring` stuff. I will merge when @HechtiDerLachs gives his O.K.

Thank you @simonbrandhorst . I agree.

HereAround commented:
> Tests fail.

Investigating this right now.

HereAround committed:
Schemes: Improve documentation and code

(*) Fix typo (lay -> lie)
(*) Remove inclusion_map
(*) Remove duplicate comments
(*) Fix mistake in documentation of hypersurface_complement
(*) Reverse roles of X and Y in examples

HereAround commented:
@simonbrandhorst @afkafkafk13 @HechtiDerLachs 

I have just cleaned up the history. The 8 commits are as follows:
1. Add test file (so that all tests of schemes can be more easily executed for debugging).
2. Move the affine schemes code to the main coding source of Oscar.
3. Spread out this code for better readability.
4. Make a couple of minor code improvements.
5. Add documentation for affine schemes.
6. Improvements to documentation as suggested by @simonbrandhorst .
7. Improvements to documentation as suggested by @afkafkafk13 .
8. Improvements to documentation by @HechtiDerLachs .

I believe this makes sense. Also, I anticipate that all tests (but the nightly builds) go green.

Consequently, I now enable the automerge. This means, that this PR will be automatically (rebased and) merged once the required tests are passed.

HechtiDerLachs committed:
Schemes: Improve documentation

fieker commented:
On Wed, Nov 02, 2022 at 03:09:25AM -0700, Anne Frühbis-Krüger wrote:
> @afkafkafk13 commented on this pull request.
> 
> 
> 
> > +julia> R, (x, y) = PolynomialRing(QQ, ["x", "y"]);
> +
> +julia> I = ideal(R, [x]);
> +
> +julia> Spec(R, I)
> +Spec of Quotient of Multivariate Polynomial Ring in x, y over Rational Field by ideal(x)
> +```
> +"""
> +Spec(R::MPolyRing, I::MPolyIdeal) = Spec(quo(R, I)[1])
> +
> +
> ***@***.*** Markdown.doc"""
> +    Spec(R::MPolyRing, U::AbsMPolyMultSet)
> +
> +Given a polynomial ring ``R`, we can localize that polynomial
> +ring at a multiplicatively closed subset ``U`` of ``R`. The spectrum
> 
> It also makes sense to me, I just wanted to point out that in Textbooks one often finds S^-1 R for the localization.
> 
We discussed this - and it would be possible. Think julia
  S^-1 = inv(S) would be a new object (type) that "only" "allows"
multipliaction by R...
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/oscar-system/Oscar.jl/pull/1652#discussion_r1011508901
> You are receiving this because you are subscribed to this thread.
> 
> Message ID: ***@***.***>

HereAround commented:
The points left open in this PR are summarized in https://github.com/oscar-system/Oscar.jl/issues/1657.
